src/mlpack/tests/cv_test.cpp

```diff
@@ -164,7 +165,22 @@ BOOST_AUTO_TEST_CASE(MSETest)
 
   BOOST_REQUIRE_CLOSE(MSE::Evaluate(lr, data, responses), expectedMSE, 1e-5);
 }
-
+/**
+* Test the Structural Similarity metric (SSIMIndex).
+*/
+BOOST_AUTO_TEST_CASE(SSIMIndexTest)
+{
+  arma::mat reference1(11, 11, arma::fill::ones);
+  arma::mat image1(11, 11);
+  image1.fill(0.8);
+  double expectedSSIM = 0.9518143962;
```

Comment:
Just wondering where does this value come from is it manually calculated or is this the output from another implementation?

Reply:
It is manually done. I used [this](http://dev.theomader.com/gaussian-kernel-calculator/) for the Gaussian window and then another site for calculating the convolution. Since the image is a single color, the mean and variance of the convoluted matrix are easily found.

Reply:
Thanks for the clarification, do you think we should add another test that calculates SSIM for an actual image?

Reply:
Hi @zoq sorry for replying back so late. I took two Einstein images and found their ssim using matlab, the value came out to be 0.703, when I did it with my implementation, it came out to be in the neighborhood of 0.9, then I printed my Gaussian matrix and found out it was not implemented correctly, only the 4th quadrant of the matrix was right. It gave a right answer for the constant color because, all quadrants are basically symmetric and thus the mean did not change (a lesson learnt). So I corrected that mistake and now the value was around 0.711. Then I saw a matlab implementation of metric, it had `filter2(window, img1, 'valid')` as the convolution step, this converted a `256x256` matrix to `246x246`, but there was no `"valid"` argument for `conv2` in armadillo, so I used `"same"` argument and then `shed_rows` and `shed_cols` functions to drop 5 rows and cols from each side but then I get 0.693. After looking at the first element of the matlab and my version of the matrix, I see that the values differ by approximately 0.3. I am still trying to figure out what to do.

Reply:
Thanks for the update, can you modify the matlab version to use `same` instead of `valid`?

Reply:
The matlab implementation with `same` gives 0.7116 and with `valid` gave 0.693. The in-built
implementation in matlab gave 0.703.

Reply:
The matlab implementation
`img1 = imread('ssim-demo-processed.pgm');
img2 = imread('ssim-demo.pgm');


[M N] = size(img1);

   if ((M < 11) | (N < 11))
     ssim_index = -Inf;
     ssim_map = -Inf;
      return
   end
   window = fspecial('gaussian', 11, 1.5);  %
   K(1) = 0.01;          % default settings
   K(2) = 0.03;          %
   L = 255;

img1 = double(img1);
img2 = double(img2);

% automatic downsampling
f = max(1,round(min(M,N)/256));
%downsampling by f
%use a simple low-pass filter 
if(f>1)
    lpf = ones(f,f);
    lpf = lpf/sum(lpf(:));
    img1 = imfilter(img1,lpf,'symmetric','same');
    img2 = imfilter(img2,lpf,'symmetric','same');

      img1 = img1(1:f:end,1:f:end);
      img2 = img2(1:f:end,1:f:end);
 end

C1 = (K(1)*L)^2;
C2 = (K(2)*L)^2;
window = window/sum(sum(window));

mu1   = filter2(window, img1, 'same');
mu2   = filter2(window, img2, 'same');
mu1_sq = mu1.*mu1;
mu2_sq = mu2.*mu2;
mu1_mu2 = mu1.*mu2;
sigma1_sq = filter2(window, img1.*img1, 'same') - mu1_sq;
sigma2_sq = filter2(window, img2.*img2, 'same') - mu2_sq;
sigma12 = filter2(window, img1.*img2, 'same') - mu1_mu2;

if (C1 > 0 & C2 > 0)
   ssim_map = ((2*mu1_mu2 + C1).*(2*sigma12 + C2))./((mu1_sq + mu2_sq + C1).*(sigma1_sq + sigma2_sq + C2));
else
   numerator1 = 2*mu1_mu2 + C1;
   numerator2 = 2*sigma12 + C2;
  denominator1 = mu1_sq + mu2_sq + C1;
   denominator2 = sigma1_sq + sigma2_sq + C2;
   ssim_map = ones(size(mu1));
   index = (denominator1.*denominator2 > 0);
   ssim_map(index) = (numerator1(index).*numerator2(index))./(denominator1(index).*denominator2(index));
   index = (denominator1 ~= 0) & (denominator2 == 0);
   ssim_map(index) = numerator1(index)./denominator1(index);
end

digitsOld = digits(10);
mssim = mean2(ssim_map);
vpa(mssim)`

Reply:
Hi @zoq, I tried figuring out what went wrong and [this](https://github.com/scikit-image/scikit-image/issues/3942#issuecomment-497692455) thread seemed to be helpful. They said that it was a matter of cropping the ssim map from the edges but it means the same as same and valid in Matlab. They also have used gaussian matrix with truncate=3.5 but I am not sure what that means. Should I close the PR as this does not appear to be progressing?